cf-agent/verify_files_utils.c

```diff
@@ -854,7 +863,16 @@ static PromiseResult SourceSearchAndCopy(EvalContext *ctx, const char *from, cha
             if (cf_stat(newfrom, &sb, attr.copy, conn) == -1)
             {
                 Log(LOG_LEVEL_VERBOSE, "Can't stat '%s'. (cf_stat: %s)", newfrom, GetErrorStr());
-                continue;                                       /* TODO FAIL? */
+                if (conn->conn_info->is_timeouted)
+                {
+                    cfPS(ctx, LOG_LEVEL_INFO, PROMISE_RESULT_INTERRUPTED, pp,
+                         attr, "connection timeout");
+                    return PROMISE_RESULT_INTERRUPTED;
```

Comment:
> There might be some other cases of this to fail.

True, `cf_stat()` can fail in a bunch of ways. But one possible cause might be network error; in that case, you don't need to differentiate with timeout: **any** network failure should terminate the promise, not only timeout.

This might help untangling this mess, since the whole "return the network timeout from deep inside the network path" issue won't even be needed to be resolved.